### ISSUE TITLE
CSV loading API limitations

### ISSUE THREAD

**Issue by fge** [jOOQ/jOOQ#4213]
I have noticed a few troubles with the CSV loading API. 

Starting with the documentation since that is the "easy" part. [The CSV example](http://www.jooq.org/doc/3.5/manual-single-page/#importing-csv) given does not actually work:

```
ID;AUTHOR_ID;TITLE
1;1;1984
2;1;Animal Farm
```

First of all, the default separator is the comma; and then the bundled CSV API requires that all fields be quoted (I've been bitten badly by that one and had to look at the postgres logs + javadoc to figure out the problem).

Then there is the problem of the escaping character; I'm not sure what is to be escaped here: is that the quoting charater only? And the API doesn't offer any means to set it...

**Comment by lukaseder** [comment]
Thanks for reporting, and I'm sorry that this has caused confusion for you.

1. You're right, we shouldn't use the semicolon in the examples, as the default is indeed a comma.
2. This:

    > and then the bundled CSV API requires that all fields be quoted

    are you sure? Our integration tests mix quoted and unquoted values... Or are you talking about field names?

> Then there is the problem of the escaping character; I'm not sure what is to be escaped here: is that the quoting charater only? And the API doesn't offer any means to set it...

Do you mean when you want to escape the quote character? RFC 4180 specifies that the double quote character needs to be repeated within a quoted string: https://tools.ietf.org/html/rfc4180#page-2 There is no separate quoting mechanism in CSV, although some libraries still offer a way to do that.

What kind of "CSV" are you importing?

**Comment by fge** [comment]
> are you sure? Our integration tests mix quoted and unquoted values... Or are you talking about field names?

When I tried and inserted without quotes it failed to insert even a single row :(

In the posgresql logs I saw this (which is quite strange to begin with):

```
2015-04-20 13:04:03 CEST [10053-1] grappa@grappadb ERROR:  null value in column "id" violates not-null constraint
2015-04-20 13:04:03 CEST [10053-2] grappa@grappadb DETAIL:  Failing row contains (null, null, null, null, null).
2015-04-20 13:04:03 CEST [10053-3] grappa@grappadb STATEMENT:  insert into "public"."matchers" ("parse_info_id") values ($1)
```

which is quite strange to begin with since the code to insert mentions several fields (details in [this SO question](http://stackoverflow.com/q/29749615/1093528).

> What kind of "CSV" are you importing?

Details are also in the question; also, it should be known that for the matchers CSV, I always split in 4 since the last column may contain any character...

Basically, a line can look like this:

> 328;SequenceMatcher;COMPOSITE;token

But in order to import into PostgreSQL I have to prepend a UUID to each row as well! Which is why I currently write dedicated CSVs.

I guess it'd be nice if there were some loading API being able to take an `Iterator<String[]>` as an argument or something like that :p

**Comment by lukaseder** [comment]
(we'll deal with the comma vs. semi-colon thing in a different issue to keep discussions on track: #4217)

**Comment by lukaseder** [comment]
> When I tried and inserted without quotes it failed to insert even a single row :(

I'm sorry, but I cannot reproduce this. An example of what we're importing in integration tests is this:

```java
String csv = "ID,First Qualifier,Last Qualifier\r" +
        "8,Hermann,Hesse\n" +
        "9,Max,Frisch\n" +
        "10,Friedrich,Dürrenmatt";
```

As you can see, there are no quotes in this example... Could you show an example where loading fails?

> I guess it'd be nice if there were some loading API being able to take an Iterator<String[]> as an argument or something like that :p

Good idea! Coincidentally, I've thought of a similar thing just recently: #4212. I've created #4218 to keep track of your idea.

**Comment by fge** [comment]
Uhmwell, I also have no headers in the CSV, maybe that explains it? Although I doubt it.

Unfortunately I have to work on something else now, and will only get back to this problem when 3.6 will be out, so I guess this issue can be closed.

One thing however about the loading API; my suggestion was to load from "CSV lines" in fact. In my process the `Reader` is not issued from the on-disk filesystem but from a `Path` opened directly from a zip file. What I had in mine was to have something like `.loadCsv(Iterator<String>)` or `.loadCsv(Iterable<String>)`.

Ahwell, I don't know. Giving too many options in the API is not a good idea either, I guess I'll adapt... Right now in fact I "insert" a `Function<String, ? extends Record>` to convert a CSV line into a record when I insert into H2, primarily for user feedback (which 3.6 will solve anyway).

**Comment by lukaseder** [comment]
> Uhmwell, I also have no headers in the CSV, maybe that explains it? Although I doubt it.

By default, the loader API ignores the first row. See [`LoaderCSVOptionsStep.ignoreRows()`](http://www.jooq.org/javadoc/latest/org/jooq/LoaderCSVOptionsStep.html#ignoreRows-int-). It's good to know, but I also doubt that this is the issue here.

> What I had in mine was to have something like .loadCsv(Iterator<String>) or .loadCsv(Iterable<String>).

I see. Yes, that makes sense as well, although, it might be a bit redundant with [`LoaderSourceStep.loadCSV(InputStream)`](http://www.jooq.org/javadoc/latest/org/jooq/LoaderSourceStep.html#loadCSV-java.io.InputStream-) and [`loadCSV(Reader)`](http://www.jooq.org/javadoc/latest/org/jooq/LoaderSourceStep.html#loadCSV-java.io.Reader-), perhaps? I haven't used the nio packages extensively, yet, but I'm not too big of a fan of the `??<String>` representation of files and when CSV is involved, the `??<String>` type might be even more confusing as it doesn't clearly communicate the fact that the lines have been split, but not the columns, for instance...

**Comment by fge** [comment]
> By default, the loader API ignores the first row. See LoaderCSVOptionsStep.ignoreRows(). It's good to know, but I also doubt that this is the issue here.

Oh f*. I didn't read this part! That may explain a lot of things! I'll keep that in mind :(

**Comment by lukaseder** [comment]
We'll add a remark...

The rationale of the `Loader` API is that it performs the inverse of the `Result.formatCSV()` API, which includes a header, much like it imports the same JSON format that is exported by `Result.formatJSON()`. That might've been a bit too clever, though...

**Comment by fge** [comment]
> Yes, that makes sense as well, although, it might be a bit redundant with LoaderSourceStep.loadCSV(InputStream) and loadCSV(Reader), perhaps?

Well, in fact just forget about that; what I do at the moment is turning a CSV line into a `? extends Record` so an `Iterator<Record>` fits the bill perfectly; and it is already scheduled, so...

**Comment by lukaseder** [comment]
Solving this issue as "works for me". All the points have been addressed in other issues
